Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-93.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: 5c544add44c8d1e4e3801204faa5cfd27e9a8131
Containerized:False
Number of Error(s): 4
/home/centos/smoketest/PR-15909/HPCC-Platform/common/remote/hooks/git/gitfile.cpp:311:43: error: \xe2\x80\x98strnstr\xe2\x80\x99 was not declared in this scope
make[2]: *** [common/remote/hooks/git/CMakeFiles/gitfile.dir/build.make:76: common/remote/hooks/git/CMakeFiles/gitfile.dir/gitfile.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:9346: common/remote/hooks/git/CMakeFiles/gitfile.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2022-03-17_11-34-22
time logs started
Prep time   :    25 sec (00:00:25)
Build time  :   132 sec (00:02:12)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   157 sec (00:02:37)
time logs end

[Build error:](https://gist.github.com/8431a2a0a3b4a35c3b94ecb1124f6553)
Build log [Build log](https://gist.github.com/HPCCSmoketest/8431a2a0a3b4a35c3b94ecb1124f6553#file-result-22-03-17-11-31-29-log)

--- a/common/remote/hooks/git/gitfile.cpp
+++ b/common/remote/hooks/git/gitfile.cpp
@@ -150,11 +150,19 @@ protected:
     git_tree * gitRoot = nullptr;
 };
 
+#define LFSsig "version https://git-lfs.github.com/spec/"
+const unsigned LFSsiglen = strlen(LFSsig);
+constexpr unsigned MIN_LFS_POINTER_SIZE=120;
+constexpr unsigned MAX_LFS_POINTER_SIZE=150;
 
 class GitRepositoryFileIO : implements CSimpleInterfaceOf<IFileIO>
 {
+    bool isLFSfile()
+    {
+        return buf.length() > LFSsiglen && memcmp(buf.toByteArray(), LFSsig, LFSsiglen)==0;
+    }
 public:
-    GitRepositoryFileIO(GitCommitTree * commitTree, const git_oid * oid)
+    GitRepositoryFileIO(GitCommitTree * commitTree, const char *gitDirectory, const git_oid * oid)
     {
         git_blob *blob = nullptr;
         int error = git_blob_lookup(&blob, git_tree_owner(commitTree->queryTree()), oid);
@@ -164,8 +172,26 @@ public:
         git_object_size_t blobsize = git_blob_rawsize(blob);
         const void * data = git_blob_rawcontent(blob);
         buf.append(blobsize, data);
-
         git_blob_free(blob);
+        if (isLFSfile())
+        {
+            Owned<IPipeProcess> pipe = createPipeProcess();
+            if (pipe->run("git-lfs", "git-lfs smudge", gitDirectory, true, true, false, 0))
+            {
+                pipe->write(blobsize, buf.toByteArray());
+                pipe->closeInput();
+                buf.clear();
+                Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
+                readSimpleStream(buf, *pipeReader);
+                pipe->closeOutput();
+            }
+            int retcode = pipe->wait();
+            if (retcode)
+            {
+                buf.clear();  // Can't rely on destructor to clean this for me
+                throw MakeStringException(0, "git-lfs returned exit status %d", retcode);
+            }
+        }
     }
     virtual size32_t read(offset_t pos, size32_t len, void * data)
     {
@@ -275,6 +301,19 @@ public:
                 throw MakeStringException(0, "git git_blob_lookup returned exit status %d", error);
 
             fileSize = git_blob_rawsize(blob);
+            if (fileSize >= MIN_LFS_POINTER_SIZE && fileSize <= MAX_LFS_POINTER_SIZE)
+            {
+                MemoryBuffer buf;
+                const void * data = git_blob_rawcontent(blob);
+                buf.append(fileSize, data);
+                if (memcmp(buf.toByteArray(), LFSsig, LFSsiglen)==0)
+                {
+                    buf.append('\0');
+                    const char *sizeptr = strstr(buf.toByteArray(), "size ");
+                    if (sizeptr)
+                        fileSize = atoi64(sizeptr+5);
+                }
+            }
             git_blob_free(blob);
         }
         return fileSize;
@@ -302,12 +341,12 @@ public:
     virtual IFileIO * open(IFOmode mode, IFEflags extraFlags) override
     {
         assertex(mode==IFOread && isExisting && !isDir);
-        return new GitRepositoryFileIO(commitTree, &oid);
+        return new GitRepositoryFileIO(commitTree, gitDirectory, &oid);
     }
     virtual IFileIO * openShared(IFOmode mode, IFSHmode shmode, IFEflags extraFlags) override
     {
         assertex(mode==IFOread && isExisting && !isDir);
-        return new GitRepositoryFileIO(commitTree, &oid);
+        return new GitRepositoryFileIO(commitTree, gitDirectory, &oid);
     }
 
 

